crates/wasm/src/wasm_planner.rs

```diff
@@ -0,0 +1,221 @@
+use crate::error::WasmResult;
+use crate::planner::Planner;
+use crate::storage::IndexedDBStorage;
+use crate::swap_record::SwapRecord;
+use anyhow::Result;
+use ark_ff::UniformRand;
+use decaf377::Fq;
+use penumbra_dex::swap_claim::SwapClaimPlan;
+use penumbra_proto::core::chain::v1alpha1::{ChainParameters, FmdParameters};
+use penumbra_proto::core::crypto::v1alpha1::{Address, DenomMetadata, Fee, StateCommitment, Value};
+use penumbra_proto::core::transaction::v1alpha1::{MemoPlaintext, TransactionPlan};
+use penumbra_proto::DomainType;
+use rand_core::OsRng;
+use serde_wasm_bindgen::Error;
+use wasm_bindgen::prelude::wasm_bindgen;
+use wasm_bindgen::JsValue;
+
+#[wasm_bindgen]
+pub struct WasmPlanner {
+    planner: Planner<OsRng>,
+    storage: IndexedDBStorage,
+}
+
+#[wasm_bindgen]
+impl WasmPlanner {
+    #[wasm_bindgen(constructor)]
+    pub async fn new() -> Result<WasmPlanner, Error> {
+        let planner = WasmPlanner {
+            planner: Planner::new(OsRng),
+            storage: IndexedDBStorage::new().await?,
+        };
+        Ok(planner)
+    }
+
+    /// Add expiry height to plan
+    /// Arguments:
+    ///     expiry_height: `u64`
+    #[wasm_bindgen(js_name = expiryHeight)]
+    pub fn expiry_height(&mut self, expiry_height: JsValue) -> Result<(), Error> {
+        self.planner
+            .expiry_height(serde_wasm_bindgen::from_value(expiry_height)?);
+        Ok(())
+    }
+
+    /// Add memo to plan
+    /// Arguments:
+    ///     memo: `MemoPlaintext`
+    pub fn memo(&mut self, memo: JsValue) -> Result<(), Error> {
+        self.memo_inner(memo)?;
```

Comment:
Same feedback as the others. I'm not sure the point of having `inner` helper functions here given the body of the helper function is so small.

Reply:
This is only necessary in order to use WasmResult<()> in the helper function

Reply:
In that case we should do this:

```rust
// In wasm_planner.rs
pub fn memo(&mut self, memo: JsValue) -> WasmResult<()> {
    let memo_proto: MemoPlaintext = serde_wasm_bindgen::from_value(memo)?;
    let _ = self.planner.memo(memo_proto.try_into()?);
    Ok(())
}

// in error.rs
impl From<WasmError> for JsValue {
    fn from(error: WasmError) -> Self {
        JsError::from(error).into()
    }
}
```

And then we can use WasmResult everywhere, even the `#[wasm_bindgen]` top level functions. It will remove the need of using the `inner` pattern too. Helper functions are not bad, but we should use them when necessary and not to please the `wasm_bindgen` gods.

Reply:
I removed some of the helper functions that seemed redundant. ✅